### ISSUE TITLE
unibi_from_term called twice (via libtermkey and neovim)

### ISSUE THREAD

**Issue by blueyed** [neovim/neovim#10671]
When looking into https://github.com/neovim/neovim/issues/9071 I've found that `unibi_from_term` gets called twice usually:
https://github.com/neovim/libtermkey/blob/b7fe3af141c53cce71a244282b69ea860452120a/driver-ti.c#L413
https://github.com/neovim/neovim/blob/b92a5bc3c4bdfddfc21638556ace6ef3fd90155c/src/nvim/tui/tui.c#L230

Given that it pokes at the file system etc it might be nice if this could be done only once, e.g. by re-using the information from libtermkey, or passing the information from Neovim to libtermkey (but currently libtermkey gets it first).

**Comment by justinmk** [comment]
> it might be nice if this could be done only once, e.g. by re-using the information from libtermkey, or passing the information from Neovim to libtermkey (but currently libtermkey gets it first).

That is done in https://github.com/neovim/neovim/pull/9094 . Need to merge that PR, but first need to set up our unibilium fork .

**Comment by blueyed** [comment]
Appears to only be done for win32 there though?
(assuming you mean https://github.com/neovim/neovim/pull/9094/files#diff-9fe05c714b192fe85bb903baf3355196R54)

**Comment by justinmk** [comment]
@blueyed yes

**Comment by erw7** [comment]
#9094 is a change to make `libtermkey` use the built-in `terminfo`. Therefore, `unibi_from_term()` is still called twice. I also considered an implementation to pass `unibi_term *` to `libtermkey`, but it was not adopted due to API incompatibility.

**Comment by justinmk** [comment]
> #9094 is a change to make `libtermkey` use the built-in `terminfo`. Therefore, `unibi_from_term()` is still called twice.

Yes, but it skips most of the work ("poking at filesystem", etc.), right?

**Comment by erw7** [comment]
> Yes, but it skips most of the work ("poking at filesystem", etc.), right?

I do not understand the meaning of "porking filesystem" well, but I don't think #9014 will help this. The implementation examined in #9014 remained [here](https://github.com/neovim/libtermkey/compare/master...erw7:feature-constructor-with-unibi_term). We need an implementation that passes `unibi_term *` like that code to `libtermkey`, or an implementation where` libtermkey` returns `unibi_term *`.

**Comment by justinmk** [comment]
> > Yes, but it skips most of the work ("poking at filesystem", etc.), right?
> 
> I do not understand the meaning of "porking filesystem" well, but I don't think #9014 will help this. The implementation examined in #9014 remained [here](https://github.com/neovim/libtermkey/compare/master...erw7:feature-constructor-with-unibi_term). We need an implementation that passes `unibi_term *` like that code to `libtermkey`, or an implementation where` libtermkey` returns `unibi_term *`.

Yeah, that's unlikely to get merged into libtermkey because the change to `driver-csi.c` is awkward. I think it would work if we modify unibilium to have something like `unibi_override(unibi_term *unibi)` which forces later calls to `unibi_from_term` to always return the forced `unibi_term` object. It's somewhat ugly, but might be the least-complicated and least amount of code. https://github.com/neovim/neovim/pull/9094/files#r312457100

**Comment by zeertzjq** [comment]
Fixed by #36233
